cmd/clusterctl/client/init_test.go

```diff
@@ -88,9 +78,108 @@ func Test_clusterctlClient_InitImages(t *testing.T) {
 			},
 			wantErr: false,
 		},
+		{
+			name: "returns error when core provider name is invalid",
+			args: args{
+				coreProvider:      "some-core-provider",
+				kubeconfigContext: "mgmt-context",
+			},
+			providers: []Provider{
+				config.NewProvider("some-core-provider", "some-core-url", clusterctlv1.CoreProviderType),
+			},
+			wantErr:              true,
+			expectedErrorMessage: "name cluster-api must be used with the CoreProvider type",
```

Comment:
lol, it took me a minute to write this comment out, because I do agree with you. Usually when testing a single function I would also no "care" about the type of error being returned. But in this case (I came with the lense of someone that is looking for the first time to the code) the InitImages function relies in at least 2 other functions that have their own use cases for errors. So I when this route, because I found myself having to looks throught these functions to find a way to make the test break.
I added the assertion on the error message itself, because I wanted to make sure that these tests where failing for the correct reasons, and not just failing, specially because we are backfilling tests, and not TDDing.

This being said, if you still feel like there is no need to have this assertion I am happy to remove it 😄

Reply:
Thanks for the explanation! Yeah...this is interesting because we aren't TDDing and instead backfilling the tests. 

I think if there is potential for the given inputs in the test to produce different errors (due to the nested calls to other functions) then I guess it's fine to specify the error message _or maybe_ that might hint to constraining the inputs further.

I do agree the important thing is to ensure that the test is failing for the correct reason. I see that this error is bubbled up _through_ `getComponentsByName` which could have returned an error from parsing the provider name. But since a valid provider name is used it goes past that to `c.configClient.Providers().Get(name, providerType)` which then surfaces the error.

I guess we can't even use the `errors.Cause` for something like this.

Hmm...this is definitely tripping me up as well and making me really think about this. 

The only other concern is that I don't want other contributors to follow the pattern of performing substring matching on errors. 

I see that @fabriziopandini 👍 so I think it's fine for now.

Feel free to squash commits regardless.